api/dp-openapi.yaml

```diff
@@ -591,6 +591,354 @@ paths:
         default:
           $ref: '#/components/responses/UnexpectedError'
 
+  /projects/badpagedata:
+    get:
+      tags:
+      - project
+      description: Gets data for mark bad page
+      responses:
+        200:
+          description: Data
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/bad_page_data'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '404':
+          $ref: '#/components/responses/NotFound'
+        '429':
+          $ref: '#/components/responses/RateLimitExceeded'
+        default:
+          $ref: '#/components/responses/UnexpectedError'
+
+  /projects/{projectid}/pickersets:
+    get:
+      tags:
+      - project
+      description: Gets pickersets for the project
+      parameters:
+      - name: projectid
+        in: path
+        description: ID of project
+        required: true
+        schema:
+          type: string
+      responses:
+        200:
+          description: pickersets
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/pickersets'
+        401:
+          $ref: '#/components/responses/Unauthorized'
+        404:
+          $ref: '#/components/responses/NotFound'
+        429:
+          $ref: '#/components/responses/RateLimitExceeded'
+        default:
+          $ref: '#/components/responses/UnexpectedError'
+
+  /rounds/{roundid}:
+    get:
+      tags:
+      - project
+      description: Gets round information
+      parameters:
+      - name: roundid
+        in: path
+        description: The round identifier (as defined by array_keys($Round_for_round_id_)
+        required: true
+        schema:
+          type: string
+      responses:
+        200:
+          description: Round data
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/round_data'
+        404:
+          $ref: '#/components/responses/NotFound'
+        429:
+          $ref: '#/components/responses/RateLimitExceeded'
+        default:
+          $ref: '#/components/responses/UnexpectedError'
+
+  /proofrounds/{proofroundid}/proofprojects/{proofprojectid}/proofpages:
```

Comment:
Can you talk about why you went for a `round -> project -> page` hierarchy? It feels a bit awkward since projects shift between rounds but rounds don't shift between projects, so the most stable part of the hierarchy is the project, so: `project -> round -> page`. It could conceivably be `/projects/{projectid}/proofrounds/{proofroundid}/...`

Reply:
It seemed more like a hierarchical address if round is put first. In the case of proofreading if you choose a round you can then choose a project which is available in that round. But doing it the other way round, if you choose a project you cannot then choose which round you want to work in because the round has to be the round the project is in.
It makes the validation work in a logical sequence: first check that the user can work in the given round and then check that the project is in the round and available.
The user is really the first component in the hierarchy although not stated explicitly so if the user is not logged in it will fall at the first fence.
Although this is not an argument either way this parallels the way the site works now in that you choose a round and are presented with a list of projects. We could make another endpoint GET `/proofrounds/{proofroundid}/proofprojects` which would get a list of projects.

Reply:
There is some redundancy in how the system works now. The project state (which is used in the PI now) defines what round, if any, the project is in. The page state also contains round information which is redundant since it must agree with the round defined by the project state. We could use project state in this API instead of round which is how the present system works; it does not matter much.
I do not think it matters that this is different to the  API `v1/projects/:projectid/pages/:pagename/pagerounds/:pageroundid` because they are doing different things. That endpoint is for looking at data and is not concerned with the user's permissions or the project state or page state. For proofreading we have to check all of these which can be done conveniently by the validators. These validators do more checks than the validators for the above endpoint.
Doing the validation in this way means that the _Page_require() functions in DPage are not needed for proofreading operations because we have already checked that everything is as it has to be for the operation we are doing.

Reply:
But nothing is stopping you from doing that enhanced validation on non-GET requests using the same URLs -- you don't need separate URLs for that.

As a user I'm looking at the new API endpoints and it seems like you're adding a lot of complexity for no discernible benefit.

Reply:
Yes, that should be possible, I'll try it out. I think it would be desirable to do the enhanced validation also on GET requests; there might be a way.

Reply:
What validations are done is determined by the names of the parts of the URL. This seems to me to be a simple way of arranging appropriate validation. I do not see how it adds any complexity.

Reply:
You're adding an entirely new set of endpoints which require new code, new documentation in the Swagger, and new sets of validators for resources that already have existing endpoints -- that's complexity.

Reply:
Being able to validate based on the name of each part of the url is one of the good parts of this system so I would be reluctant to reject it. Since we have to validate anyway this seems like the simplest and best way to do it.

Reply:
But you can do that already...
```diff
diff --git a/api/v1_projects.inc b/api/v1_projects.inc
index c2dda153d..b77b74448 100644
--- a/api/v1_projects.inc
+++ b/api/v1_projects.inc
@@ -393,6 +393,26 @@ function api_v1_project_pagedetails($method, $data, $query_params)

 function api_v1_project_page_round($method, $data, $query_params)
 {
+    // handle GET and POST requests
+    if ($method == "GET") {
+        // everyone can get a page
+        return _api_v1_project_page_round_GET($data);
+    } elseif ($method == "POST") {
+        $project = $data[":projectid"];
+        $round = $data[":roundid"];
+        $page_name = $data[":pagename"];
+
+        // pages can only be modified if the project is in the correct
+        // round/state and the user has the page checked out, etc
+        if (! $project->state ... && ...) {
+            throw new UnauthorizedError("You are not authorized to change the page at this time.");
+        }
+
+        // update page
+    }
+}
+
+function _api_v1_project_page_round_GET($data) {
     if ($data[":pageroundid"] == "OCR") {
         $text_column = "master_text";
         $user_column = "NULL";
```

Reply:
Although this is not the main issue I'll mention it now in case I forget: there is a case where not doing any validation before getting a page could be a problem: when the user reverts to the last save. If the project had moved to a different state he would not become aware of it.

Reply:
If we used round rather than pageround for proofreading purposes then the GET case could validate also.

Reply:
> when the user reverts to the last save. If the project had moved to a different state he would not become aware of it.

This should be handled by the client. The data returned from the GET has the data needed for the client to validate this -- this isn't the responsibility of the endpoint, because it isn't an authorization issue.

Reply:
You could look at it like that and I'm quite happy to go along with it but there would have to be two requests instead of one. Can we not make the api requests mean whatever we want them to? Like "Get this page for me to proofread" and the server could reply "No, this page is no longer available to you for proofreading" (though you can look at it).

Reply:
Proofreading is just one of many use-cases for the API. Right now any user can view any page text for any page in any project in any round the project has been in so far regardless of the page's save status (not checked out, checked out, returned, etc) -- that is: they are authorized to do so. Why would we change that just to satisfy one specific use-case?

We can make the API requests mean whatever we want to, but not at the expensive of reducing the functionality of the API just because it makes it easier for one specific use case.

Reply:
Good, I think we can proceed on this basis. I'll revise the doc.